### PR TITLE
Revisions #4: Add revisions list and revisions diff viewer

### PR DESCRIPTION
**Pull request by bperson**
~**Note: This builds on top of #14242, so it's not merge-able yet**~

4th chunk coming from my WIP PR to introduce post revisions in calypso: #13367, this is mostly UI code to introduce a list of revisions in the post editor sidebar, a diff viewer replacing the main editor with a diff view of the revision currently selected and some code to enable "loading" a revision in the editor.

### testing

- Tap on "x revisions" in the post editor sidebar (shown only on posts with multiple revisions)
- It should automatically fetch the revisions for the post you're currently editing
- On desktop, once the revisions are fetched, it should automatically display the latest revision in the diff viewer
- On mobile, once the revisions are fetched, it should NOT automatically select a revision.
- On mobile, selecting a revision should switch focus to the diff viewer, you can then reopen the revisions sidebar by tap-ing the "history" icon in the editor ground control.
- Once you have a revision selected in the list, you can "load the revision in the editor", on mobile, this should dismiss the entire sidebar.

### screenshots

desktop

mobile

### CODE REVIEW

**Comment by bperson** [comment]
Now that #14943 is merged, CircleCI is green, it's also rebased on the latest master

**Comment by bperson** [comment]
#### Updates
- added a couple screenshots
- updated `EditorRevisionsList` to get `siteId` and `postId` from Redux instead of having to pass down props (coming from a review on the follow-up PR: #14928 )

**Comment by folletto** [comment]
cap-drop

I did a design review. Looks REALLY good so far. I have some minor notes:

1. Text seem using a light gray. Switch "X ago by Name" to `$gray-dark`.
2. Some blocks in the list have no subtitle (`editor-revisions-list__changes`). Why is that? Can we make sure it's always present?
3. It has some odd elements inside on mobile ([screenshot](https://user-images.githubusercontent.com/4389/27376413-5813315c-5672-11e7-9781-16ad442ffbf8.png), also the author in the top screenshot above). Can we make sure they don't pop-in on mobile?

I'm sure we can iterate and improve this further, but I think it's already a huge improvement, so let's try to ship this. :)

**Comment by bperson** [comment]
> Text seem using a light gray. Switch "X ago by Name" to $gray-dark.

Updated, should we keep the hover color coming from using `Button`s though? (we still have the background color changing on hover right now).

> Some blocks in the list have no subtitle (editor-revisions-list__changes). Why is that? Can we make sure it's always present?

AFAIK, this should only happen when there is no "word changes" detected between 2 revisions, for instance if you change a number or punctuation ([which aren't counted as words by `countWords`](https://github.com/Automattic/wp-calypso/blob/master/client/lib/text-utils/index.js#L23)).

I'm not sure whether it's worth it to introduce a "less-strict" count words for revisions (making it inconsistent with the word count at the bottom of the editor) though?

screen shot 2017-06-21 at 12 01 28

> It has some odd elements inside on mobile (screenshot, also the author in the top screenshot above). Can we make sure they don't pop-in on mobile?

Missed it during the split from the WIP PR :/, it's fixed:

screen shot 2017-06-21 at 12 09 36

**Comment by folletto** [comment]
> Updated, should we keep the hover color coming from using Buttons though? (we still have the background color changing on hover right now).

Are we sure inheriting from buttons is a good idea?

Anyhow, the design is meant to be like the accordions styling:

screen shot 2017-06-21 at 12 38 55

Notice the hover turns the text blue. Sorry, I recognize it wasn't evident from the mockups.


> AFAIK, this should only happen when there is no "word changes" detected between 2 revisions, for instance if you change a number or punctuation

Odd. I had it on the first revision too, which was fairly big. If it's "smaller" we could just say "Minor changes" (in `$gray`). Mostly because it breaks the rhythm to have some blocks smaller (for this kind of design, different designs might build on that as a feature).

**Comment by bperson** [comment]
> Are we sure inheriting from buttons is a good idea?

probably not, considering how much we're "tweaking" it, I've updated it and the rest of the CSS to more closely match what can be found in an accordion:

2 nitpick questions:
- with the accordion spacing, the subtitle (the changes summary) seems a bit close to the title ("x ago by who")? Maybe the "g" (from "ago") and the "d" (from "word") making it weird?
- in the accordion, the subtitle is [in `$gray-text-min`](https://github.com/Automattic/wp-calypso/blob/master/client/components/accordion/style.scss#L109), should I also use this? (instead of `$gray`)

screen shot 2017-06-21 at 14 49 03

> Odd. I had it on the first revision too, which was fairly big. If it's "smaller" we could just say "Minor changes" (in $gray). Mostly because it breaks the rhythm to have some blocks smaller (for this kind of design, different designs might build on that as a feature).

Oops, my bad, it will also occur on the first revisions, right now [it's defaulting to no change](https://github.com/Automattic/wp-calypso/blob/master/client/state/data-layer/wpcom/posts/revisions/index.js#L75) but we can very easily change that to consider that every word is actually "added" from what essentially was a blank page before. Since I would like to keep this PR on UI code alone, I'll cut another PR to fix it.

I've also added "minor changes" for all cases where we don't have "word diff": 

screen shot 2017-06-21 at 14 48 21

**Comment by folletto** [comment]
> with the accordion spacing, the subtitle (the changes summary) seems a bit close to the title ("x ago by who")? Maybe the "g" (from "ago") and the "d" (from "word") making it weird?

Yes, good point. Try adding some to your discretion and let's review seeing the screenshots :)

> in the accordion, the subtitle is in `$gray-text-min`, should I also use this?

Good point too, let's do that.

> it will also occur on the first revisions, right now it's defaulting to no change but we can very easily change that to consider that every word is actually "added" from what essentially was a blank page before. Since I would like to keep this PR on UI code alone, I'll cut another PR to fix it.

Sounds good. It's valuable as sometimes the bulk is added immediately and sometimes it's jut a few words. :)

> I've also added "minor changes" for all cases where we don't have "word diff":

Excellent. 👍

**Comment by bperson** [comment]
5px of padding:

screen shot 2017-06-21 at 15 13 22

> Good point too, let's do that.

updated :ok_hand:

**Comment by folletto** [comment]
Cool! 
Design wise we're good 👍

**Comment by bperson** [comment]
Awesome!, thanks a lot for the review @folletto, I've created #15354 to fix the issue with the first revision change summary.

It's time to find some code reviewers now ;)

File: client/post-editor/editor-diff-viewer/index.jsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	selectedRevisionId: PropTypes.number,
+	siteId: PropTypes.number,
+};
+
+export default connect(
+	( state, ownProps ) => ( {
+		contentChanges: getPostRevisionChanges( state, ownProps.siteId, ownProps.postId, ownProps.selectedRevisionId ),
+		revision: normalizeForEditing(
+			getPostRevision( state, ownProps.siteId, ownProps.postId, ownProps.selectedRevisionId )
+		),
```
Comment (mcsf) [inline client/post-editor/editor-diff-viewer/index.jsx]: Just like `getPostRevision`, `normalizeForEditing` should probably be cached via `createSelector`. This would avoid the cost of deeply cloning revision objects, and it would unlock `connect`'s `pure` behavior.

File: client/post-editor/editor-revisions-list/header.jsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+		>
+			{ translate( 'Load revision in the editor' ) }
+		</Button>
+	</div>
+);
+
+EditorRevisionsListHeader.propTypes = {
+	loadRevision: PropTypes.func,
+	selectedRevisionId: PropTypes.number,
+	translate: PropTypes.func,
```
Comment (mcsf) [inline client/post-editor/editor-revisions-list/header.jsx]: Since the component calls `this.props.translate` as a function, I suggest making this a required prop. See if the other props need the same.

File: client/post-editor/editor-revisions-list/index.jsx [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+		return {
+			postId,
+			revisions: orderBy(
+				map(
+					getPostRevisions( state, siteId, postId ),
+					normalizeForDisplay
+				),
+				'date',
+				'desc'
+			),
```
Comment (mcsf) [inline client/post-editor/editor-revisions-list/index.jsx]: Same as previously: for perf + leveraging immutability, make this a memoized selector.

File: client/post-editor/editor-diff-viewer/index.jsx (hunk 1 above)
Comment (bperson) [inline client/post-editor/editor-diff-viewer/index.jsx]: yep already thought about that, I'm not too sure how to tackle it tbh: I can't use `normalizeForEditing` as a parameter to the selector since a function is an [invalid arg type](https://github.com/Automattic/wp-calypso/blob/master/client/lib/create-selector/index.js#L17). 4 solutions I thought about:

1- Demultyply selectors: `getPostRevisionForEditing`, `getPostRevisionsForEditing`, `getPostRevisionForDisplay` and `getPostRevisionsForDisplay`.
2- Have an internal mapping `string` <-> `function` in the selectors so that we pass a "normalizer" string: `getPostRevision( state, siteId, postId, revisionId, 'editing')` and internally, it will call `normalizeForEditing` (`display` for `normalizeForDisplay`)
3- Do 1- but refactor `getPostRevision` and `getPostRevisions` under the same selector to limit the "API" spread of demultiplying selectors like in 1-? This will make selector usage intent a bit less clear, as sometimes we will do `get(getPostRevisions( state, siteId, postId, revisionId ), 0, null)` using plural and returning an array even though we already know we only want one revision :/ 
4- Find a way to accept (and 'freeze"?) functions as args for the selectors using `createSelector`? <-- I haven't dug much in that direction tbh

File: client/post-editor/editor-revisions-list/item.jsx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import React, { PureComponent, PropTypes } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import PostTime from 'reader/post-time';
+
+class EditorRevisionsListItem extends PureComponent {
+	selectRevision = event => {
+		const revisionId = parseInt( event.currentTarget.dataset.revisionId, 10 );
```
Comment (bperson) [inline client/post-editor/editor-revisions-list/item.jsx]: nice catch, leftover from when it wasn't a component on its own ;) fixed in the latest update

File: client/post-editor/editor-sidebar/index.jsx [hunk 5]
```diff
@@ -84,7 +91,17 @@ export default class EditorSidebar extends Component {
...
-				<SidebarRegion className="editor-sidebar__nested-region" />
+				<SidebarRegion className="editor-sidebar__nested-region">
+					{
+						nestedSidebar === NESTED_SIDEBAR_REVISIONS
+							? <EditorRevisionsList
+								loadRevision={ loadRevision }
+								selectedRevisionId={ selectedRevisionId }
+								selectRevision={ selectRevision }
+							/>
+							: null
```
Comment (bperson) [inline client/post-editor/editor-sidebar/index.jsx]: fixed in the latest update

File: client/post-editor/editor-revisions-list/header.jsx (hunk 2 above)
Comment (bperson) [inline client/post-editor/editor-revisions-list/header.jsx]: I've done a full pass on all prop types, if there is a usage where it can lead to a runtime exception (array/object access and function calls), I've `isRequired`-ify the prop type. Is there any "guideline" in calypso to choose between `isRequired` vs default props?

Comment (mcsf) [inline client/post-editor/editor-revisions-list/header.jsx]: Good of you to ask, though there isn't a guideline — you can stumble upon a `translate: func.isRequired` type just as you can stumble upon a `translate: identity` default.

File: client/post-editor/editor-diff-viewer/index.jsx (hunk 1 above)
Comment (mcsf) [inline client/post-editor/editor-diff-viewer/index.jsx]: Without giving it too much thought, I'd lean towards number 2.

**Comment by bperson** [comment]
@mcsf all should be fixed. It adds a couple of tests for the changes to the memoized selectors hence why the diff-meter is going up.

File: client/post-editor/post-editor.jsx [hunk 6]
```diff
@@ -228,9 +244,39 @@ export const PostEditor = React.createClass( {
 	},
 
 	toggleNestedSidebar: function( nestedSidebar ) {
+		if (
+			this.state.nestedSidebar === NESTED_SIDEBAR_REVISIONS &&
+			this.state.nestedSidebar !== nestedSidebar
+		) {
+			this.selectRevision( null );
```
Comment (mcsf) [inline client/post-editor/post-editor.jsx]: Seems like we're mixing concerns here, are we not? Any "extension" of the editor can make use of nested sidebars, so this logic is expected to be generic, yet there is now logic specific to revisions; in addition, it can be argued that this scatters the state-keeping logic of revisions in more places, thus hurting maintainability.

Can we find a more decoupled (or at least indirect) way to achieve the same result?

File: client/post-editor/post-editor.jsx [hunk 7]
```diff
@@ -431,19 +492,21 @@ export const PostEditor = React.createClass( {
...
-
-		this.setState( { showAutosaveDialog: false, isLoadingAutosave: true } );
-
-		edits = {
-			title: autosaveData.title,
-			excerpt: autosaveData.excerpt,
-			content: autosaveData.content
-		};
+		this.setState( { showAutosaveDialog: false } );
+		this.restoreRevision( this.state.post.meta.data.autosave );
```
Comment (mcsf) [inline client/post-editor/post-editor.jsx]: Though this access isn't new to this PR, how sure are we that this nested property can be safely accessed? Does it warrant the use of `get`?

File: client/post-editor/post-editor.jsx [hunk 8]
```diff
@@ -248,9 +264,39 @@ export const PostEditor = React.createClass( {
...
+		if (
+			selectedRevisionId !== null &&
+			isWithinBreakpoint( '<660px' )
+		) {
+			this.props.setLayoutFocus( 'content' );
+		}
+	},
+
+	loadRevision: function( revision ) {
+		this.toggleNestedSidebar( NESTED_SIDEBAR_NONE );
```
Comment (jblz) [inline client/post-editor/post-editor.jsx]: `this.toggleNestedSidebar` does not exist

Comment (mcsf) [inline client/post-editor/post-editor.jsx]: Missed it in the rebase. Fixed.

**Comment by jblz** [comment]
@mcsf Here's a demo of the issue I showed you:
https://cloudup.com/c3--PP5Q2oS

If you start typing in the content portion of the editor from the start, the draft gets saved with the initial content, a post id gets placed in the URL bar, etc, but it looks like no revision exists until a subsequent save.

This seems like an editor bug that's outside this PR & playing with this just surfaced it. I'm documenting here, so we're sure to follow up with this in a separate issue since there's a possibility of data loss since the initial contents don't make it into the list of revisions.

**Comment by folletto** [comment]
> This seems like an editor bug that's outside this PR & playing with this just surfaced it. I'm documenting here, so we're sure to follow up with this in a separate issue since there's a possibility of data loss since the initial contents don't make it into the list of revisions.

Did you already open the new issue? Seems a rough edge that as you mention might have impact even outside the revisions discussion.

**Comment by jblz** [comment]
@folletto just did #17183. Thanks.

File: client/post-editor/editor-revisions-list/header.jsx [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * Internal dependencies
+ */
+import Button from 'components/button';
+
+const EditorRevisionsListHeader = ( { loadRevision, selectedRevisionId, translate } ) => (
+	<div className="editor-revisions-list__header">
+		<Button
+			className="editor-revisions-list__load-revision"
+			compact={ true }
```
Comment (spen) [inline client/post-editor/editor-revisions-list/header.jsx]: We could lose `={ true }`, since `true` would be the implied value of `compact` without a given value.

File: client/post-editor/editor-revisions-list/index.jsx [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+		this.props.loadRevision( this.props.selectedRevision );
+	}
+
+	trySelectingRevision() {
+		if (
+			this.props.selectedRevisionId === null &&
+			this.props.revisions.length > 0 &&
+			isWithinBreakpoint( '>660px' )
+		) {
+			this.props.selectRevision( this.props.revisions[ 0 ].id );
```
Comment (spen) [inline client/post-editor/editor-revisions-list/index.jsx]: Could we use `_.first` here? 
We could call it in place or outside of the conditional to give us something like:
```js
const firstRevision = first( revisions );
if (
  ...
  !! firstRevision &&
  ...
) {
  selectRevision( firstRevision.id );
}
```

Also, will there ever chance of `.id` being undefined?

File: client/post-editor/editor-revisions-list/item.jsx [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+		return (
+			<button
+				className="editor-revisions-list__button"
+				onClick={ this.selectRevision }
+				type="button"
+			>
+				<span className="editor-revisions-list__date">
+					<PostTime date={ this.props.revision.date } />
+				</span>
+				&nbsp;
```
Comment (spen) [inline client/post-editor/editor-revisions-list/item.jsx]: Looks like the span below could come up empty if `isObject( this.props.revision.author )` is falsey.
In those cases would a `nbsp` cause some off-centring? Perhaps the `nbsp` would be better off inside of that conditional

File: client/post-editor/post-editor.jsx [hunk 12]
```diff
@@ -138,6 +141,19 @@ export const PostEditor = React.createClass( {
...
 
+	componentDidUpdate( prevProps, prevState ) {
+		if (
+			prevState.nestedSidebar !== NESTED_SIDEBAR_NONE &&
+			this.state.nestedSidebar === NESTED_SIDEBAR_NONE
+		) {
+			// NOTE: Make sure we scroll back to the top AND trigger a scroll
+			// event no matter the scroll position we're coming from.
+			// ( used to force-reset TinyMCE toolbar )
+			window.scrollTo( 0, 1 );
```
Comment (spen) [inline client/post-editor/post-editor.jsx]: smart :)

File: client/post-editor/post-editor.jsx [hunk 13]
```diff
@@ -340,56 +386,68 @@ export const PostEditor = React.createClass( {
...
-									</SegmentedControlItem>
-								</SegmentedControl>
+							<div className={ classNames(
+								'post-editor__inner-content',
+								{ 'is-shown': this.state.nestedSidebar === NESTED_SIDEBAR_NONE }
+							) }>
+								<FeaturedImage
+									site={ site }
+									post={ this.state.post }
+									maxWidth={ 1462 } />
```
Comment (spen) [inline client/post-editor/post-editor.jsx]: I see that this was existing - but should we generally use constants for values like this? 1462 looks arbitrary here - a constant could give it some meaning & context.

Comment (mcsf) [inline client/post-editor/post-editor.jsx]: Indeed, this should be better explained. I traced it back to `10818-gh-calypso-pre-oss` 😅 . As the saying goes, patches welcome :trollface: . (In any case, doesn't belong in this PR; not to mention that the way GitHub renders the diff, it'd be very hard to spot that change in here.)

File: client/post-editor/editor-revisions-list/item.jsx (hunk 11 above)
Comment (mcsf) [inline client/post-editor/editor-revisions-list/item.jsx]: Good spotting, but local testing doesn't show any centring issues. I'll leave as-is.

File: client/post-editor/editor-revisions-list/index.jsx (hunk 10 above)
Comment (mcsf) [inline client/post-editor/editor-revisions-list/index.jsx]: Very subjective, but when I applied this suggestion I didn't really like it. :) Maybe because the usage of `first` is so local, or because of the `const` that gets introduced, requiring a `!!` cast, instead of just relying on `length`.

Comment (spen) [inline client/post-editor/editor-revisions-list/index.jsx]: Fair fair :P
